### PR TITLE
Fix HttpListener test hang on ILC

### LINKED ISSUE #22110: System.Net.HttpListener tests hanging in UAPAOT
Some of the HttpListener tests are hanging in AOT mode only. The failures are intermittent and the following may not be a complete list:

- [ ] 7 tests in HttpRequestStreamTests.cs 
- [x] 2 test in HttpResponseStreamTests.cs

### PR DESCRIPTION
**Pull request by caesar1995**
In test file HttpRequestStreamTests, most of test cases do:
1. Get the HttpListenerContext: `Task<HttpListenerContext> contextTask = _listener.GetContextAsync();`.
2. Using a HttpClient to Send a POST request to the specified Uri HttpListener listening on: `Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, "hello");`.
3. Get the `HttpListenerRequest request` from HttpListenerContext.Request, and do some testings and validations.
4. Use `context.Response.Close();` to finish the operation.

The `public async Task<HttpListenerRequest> GetRequest(bool chunked)` in `GetContextHelper.cs` causes problem because:
For the `PostAsync` operation, the returned Task<HttpResponseMessage> object will complete after the whole response (including content) is read.
We use `context.Response.Close()` to send the response to the client by closing the response.

Above procedures worked well on netcore. However, on ILC, we need to explicitly `await clientTask` to finish. (First few runs are working, but if run the test case for many times, i.e a test case has many inline data, it will hang.)

Fix #22110

### CODE REVIEW

**Comment by stephentoub** [comment]
> Above procedures worked well on netcore. However, on ILC, we need to explicitly await clientTask to finish.

What explains the difference?

**Comment by CIPop** [comment]
@davidsh the change @caesar1995 made is only on the client side:
1. With this, the code will create a new HttpClient for every test instead of reusing an existing client (which is the recommended pattern)
2. awaiting the client POST operation points to a potential issue in the request object lifetime (or a TPL ILC issue which now seems less likely)

@caesar1995 can you obtain Debug.WriteLine logs from ILC (I know ETW is still broken...). It may be interesting to diff between logs with and w/o your change to better understand why the tests are hanging.

**Comment by CIPop** [comment]
> @caesar1995 can you obtain Debug.WriteLine logs from ILC (I know ETW is still broken...). It may be interesting to diff between logs with and w/o your change to better understand why the tests are hanging.

Also you'll have to `set WINHTTPHANDLER_TRACE = 1` environment variable and use [DbgView](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview).

**Comment by caesar1995** [comment]
> What explains the difference?

I don't know yet. I have studied pattern for test case that will never fail, then apply that pattern to all test cases, and find out tests are not hanging anymore. I need more investigation on why there is a difference.

> can you obtain Debug.WriteLine logs from ILC...

Yes, I can make an example to show the difference on the necessity of the final `await PostAsync` operation.

Here is the test case that I have modified:

```c#
        [Theory]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        [InlineData(false)]
        [InlineData(true)]
        public async Task CanRead_Get_ReturnsTrue(bool chunked)
        {
            Debug.WriteLine("Start");
            Task<HttpListenerContext> contextTask = _listener.GetContextAsync();
            using (HttpClient client = new HttpClient())
            {
                Debug.WriteLine("step 1");
                client.DefaultRequestHeaders.TransferEncodingChunked = chunked;
                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent("Hello"));

                Debug.WriteLine("step 2");
                HttpListenerContext context = await contextTask;
                HttpListenerRequest request = context.Request;

                using (Stream inputStream = request.InputStream)
                {
                    Assert.True(inputStream.CanRead);
                }
                
                Debug.WriteLine("step 3");
                context.Response.Close();
                //await clientTask;
            }
            Debug.WriteLine("End");
            Debug.WriteLine("");
        }
```

If we don't comment out the last `await clientTask` line, everything works, and here is the log:

good1
good2

If we comment it out, test will hang after step 2:
bad

**Comment by stephentoub** [comment]
> I don't know yet. I have studied pattern for test case that will never fail, then apply that pattern to all test cases, and find out tests are not hanging anymore. I need more investigation on why there is a difference.

Then how can we state (from the linked issue): "I can confirm that it's not a product issue"?

**Comment by caesar1995** [comment]
> Then how can we state (from the linked issue): "I can confirm that it's not a product issue"?

For this hanging issue on HttpListenerRequest, if customers are using the correct pattern, it can be avoided.

**Comment by davidsh** [comment]
>For this hanging issue on HttpListenerRequest, if customers are using the correct pattern, it can be avoided.

What makes us think this pattern is THE correct pattern?  Yes, it seems to result in the test not hanging.  But if we don't understand why it works, then we should not merge the PR.

**Comment by stephentoub** [comment]
> For this hanging issue on HttpListenerRequest, if customers are using the correct pattern, it can be avoided.

That just means we're working around a behavior difference by changing the tests to avoid a hang. That doesn't necessarily mean there isn't a problem or that we've fixed it. You might be right, there may not be a product issue, but I don't believe we know that, and the fact the tests are passing on several implementations and not on another suggests there is something lurking.

**Comment by caesar1995** [comment]
Thanks for the clarification! Now I understand the problem I need to address.

I will continue my investigation.

**Comment by davidsh** [comment]
>Also you'll have to set WINHTTPHANDLER_TRACE = 1 environment variable and use DbgView.

Turns out this won't work here.  That only works for .NET Core HttpClient that uses WinHttpHandler.

These tests are using UWP HttpClient which doesn't use WinHttpHandler.

So, this tracing won't work.

**Comment by CIPop** [comment]
@caesar1995 @davidsh Sounds like we're missing ETW/Debug tracing from the WinRT handler - adding it would help a lot in this case to identify the root cause.
Repro-ing in a 2-box environment and collecting Wireshark would also help.

**Comment by davidsh** [comment]
>@caesar1995 @davidsh Sounds like we're missing ETW/Debug tracing from the WinRT handler - adding it would help a lot in this case to identify the root cause.

While there is no specific ETW tracing in the .NET layer of UWP HttpClient, there is ETW tracing already in the WinRT Windows.Web.Http layer.  Turning that on could be useful.  That should work because it is ETW tracing in the native WinRT layer.

**Comment by davidsh** [comment]
@dotnet-bot Test NETFX x86 Release Build (build still running after 1+ day)
https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netfx+CGroup_Release+AGroup_x86+TestOuter_false_prtest/1771/

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 1]
```diff
@@ -347,53 +348,86 @@ public void Dispose()
...
 
                 context.Response.Close();
+                await clientTask;
             }
         }
 
         [Fact]
         public async Task CanSeek_Get_ReturnsFalse()
         {
-            HttpListenerRequest response = await _helper.GetRequest(chunked: true);
```
Comment (CIPop) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: @caesar1995 I don't see other references to _helper yet it's still defined - any reasons?

Comment (caesar1995) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: You are right, the ContextHelper is no longer needed for this test class. I will fix this in my new PR.

**Pull request merged**